File: src/main/java/com/dmarchesseault/tools/StopWatch.java
package com.dmarchesseault.tools;

public class StopWatch
{
//    private boolean _started = false;
    private long _timeStarted;
    private long _timeStopped;

    /**
     * Start the timer
     *
     * @return this instance
     */
    public StopWatch start()
    {
        if (_timeStarted != 0 && _timeStopped == 0)
        {
            throw new IllegalStateException("Already started");
        }

        _timeStarted = System.currentTimeMillis();
        return this;
    }

    /**
     * Stops the timer.
     *
     * @return this instance
     */
    public StopWatch stop()
    {
        if (_timeStarted == 0)
        {
            throw new IllegalStateException("Not started");
        }
        else if (_timeStopped != 0)
        {
            throw new IllegalStateException("Already stopped");
        }

        _timeStopped = System.currentTimeMillis();
        return this;
    }

    /**
     * Returns the elapsed time (ms) since start.
     *
     * @return elapsed time (ms) since start.
     */
    public long elapsed()
    {
        if (_timeStarted == 0)
        {
            return 0;
        }
        else if (_timeStopped == 0)
        {
            return System.currentTimeMillis() - _timeStarted;
        }
        else
        {
            return _timeStopped - _timeStarted;
        }

    }


}

File: sudoku-js/src/main/js/tools/StopWatch.js
console && console.log("### StopWatch.js begin");

(function(ns, lang)
{
    // Constructor
    ns.StopWatch = function()
    {
        this._timeStarted = 0;
        this._timeStopped = 0;
    };

    /**
     * Start the timer
     *
     * @return this instance
     */
    //public StopWatch start()
    ns.StopWatch.prototype.start = function()
    {
        if (this._timeStarted !== 0 && this._timeStopped === 0)
        {
            throw new ns.IllegalStateException("Already started");
        }

        this._timeStarted = Date.now();
        return this;
    };

    /**
     * Stops the timer.
     *
     * @return this instance
     */
    //public StopWatch stop()
    ns.StopWatch.prototype.stop = function()
    {
        if (this._timeStarted === 0)
        {
            throw new ns.IllegalStateException("Not started");
        }
        else if (this._timeStopped !== 0)
        {
            throw new ns.IllegalStateException("Already stopped");
        }

        this._timeStopped = Date.now();
        return this;
    };

    /**
     * Returns the elapsed time (ms) since start.
     *
     * @return elapsed time (ms) since start.
     */
    //public long elapsed()
    ns.StopWatch.prototype.elapsed = function()
    {
        if (this._timeStarted === 0)
        {
            return 0;
        }
        else if (this._timeStopped === 0)
        {
            return Date.now() - this._timeStarted;
        }
        else
        {
            return this._timeStopped - this._timeStarted;
        }

    };

    ns.StopWatch.prototype.LOG = ns.StopWatch.LOG = Logger.getLogger("StopWatch ");

}(Module("tools"),Module("lang")));

console && console.log("### StopWatch.js end");


